OsmAnd-java/src/main/java/net/osmand/router/RouteColorize.java

```diff
@@ -0,0 +1,416 @@
+                double resultAlpha = minPaletteColor.getAlpha() + percent * (maxPaletteColor.getAlpha() - minPaletteColor.getAlpha());
+                return new Color((int) resultRed, (int) resultGreen, (int) resultBlue, (int) resultAlpha);
+            }
+        }
+        return getDefaultColor();
+    }
+
+    public void setPalette(double[][] palette) {
+        this.palette = palette;
+        checkPalette();
+        sortPalette();
+    }
+
+    private Color getDefaultColor() {
+        return new Color(0, 0, 0, 0);
+    }
+
+    private List<Data> simplify() {
+        if (dataList == null) {
+            dataList = new ArrayList<>();
+            for (int i = 0; i < latitudes.length; i++) {
+                //System.out.println(latitudes[i] + " " + longitudes[i] + " " + values[i]);
+                dataList.add(new Data(i, latitudes[i], longitudes[i], values[i]));
+            }
+        }
+        List<Node> nodes = new ArrayList<>();
+        List<Node> result = new ArrayList<>();
+        for (Data data : dataList) {
+            nodes.add(new net.osmand.osm.edit.Node(data.lat, data.lon, data.id));
+        }
+        OsmMapUtils.simplifyDouglasPeucker(nodes, zoom + 5, 1, result, true);
+
+        List<Data> simplified = new ArrayList<>();
+
+        for (int i = 1; i < result.size() - 1; i++) {
+            int prevId = (int) result.get(i - 1).getId();
+            int currentId = (int) result.get(i).getId();
+            List<Data> sublist = dataList.subList(prevId, currentId);
+            simplified.addAll(getExtremums(sublist));
+        }
+        return simplified;
+    }
+
+    private List<Data> getExtremums(List<Data> subDataList) {
+        if (subDataList.size() <= 2)
+            return subDataList;
+
+        List<Data> result = new ArrayList<>();
+        double min;
+        double max;
+        min = max = subDataList.get(0).val;
+        for (Data pt : subDataList) {
+            if (min > pt.val)
+                min = pt.val;
+            if (max < pt.val)
+                max = pt.val;
+        }
+
+        double diff = max - min;
+
+        result.add(subDataList.get(0));
+        for (int i = 1; i < subDataList.size() - 1; i++) {
+            double prev = subDataList.get(i - 1).val;
+            double current = subDataList.get(i).val;
+            double next = subDataList.get(i + 1).val;
+            Data currentData = subDataList.get(i);
+
+            if ((current > prev && current > next) || (current < prev && current < next)
+                    || (current < prev && current == next) || (current == prev && current < next)
+                    || (current > prev && current == next) || (current == prev && current > next)) {
+                Data prevInResult;
+                if (result.size() > 0) {
+                    prevInResult = result.get(0);
+                    if (prevInResult.val / diff > 0.05d) {// check differences in 5%
```

Comment:
Better to create constant  (so it's obvious )
private static final double MIN_DIFFERENCE_SLOPE.... = 0.05d;